### PR TITLE
Don't add duplicate docs to case

### PR DESCRIPTION
**Pull request by rkondratowicz**
If there's already a document with the same url in CCD, don't add it.

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapper.java [hunk 1]
```diff
@@ -13,7 +18,20 @@ public SupplementaryEvidenceMapper() {
         // empty mapper construct
     }
 
-    public SupplementaryEvidence mapEnvelope(Envelope envelope) {
-        return new SupplementaryEvidence(mapDocuments(envelope.documents));
+    public SupplementaryEvidence map(List<Document> existingDocs, List<Document> envelopeDocs) {
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapper.java]: I think we should either split the mapping and combining of documents or perhaps rename this class and the method.

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/servicebus/model/Envelope.java [hunk 2]
```diff
@@ -43,8 +43,4 @@ public Envelope(
         this.documents = documents;
         this.ocrData = ocrData;
     }
-
-    public void addDocuments(List<Document> documents) {
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/servicebus/model/Envelope.java]: 👍 back to immutability

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapperTest.java [hunk 3]
```diff
@@ -18,45 +20,84 @@
...
+            .containsExactly(
+                tuple("a.pdf", "aaa", "type_a", "subtype_a", toLocalDateTime(existingDocs.get(0).scannedAt), "http://localhost/a.pdf"),
+                tuple("b.pdf", "bbb", "type_b", "subtype_b", toLocalDateTime(existingDocs.get(1).scannedAt), "http://localhost/b.pdf"),
+                tuple("x.pdf", "xxx", "type_x", "subtype_x", toLocalDateTime(envelopeDocs.get(0).scannedAt), "http://localhost/x.pdf"),
+                tuple("y.pdf", "yyy", "type_y", "subtype_y", toLocalDateTime(envelopeDocs.get(1).scannedAt), "http://localhost/y.pdf")
+            );
+    }
+
+    @Test
+    public void should_not_add_document_from_envelope_if_its_already_present() {
```
Comment (lgonczar) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapperTest.java]: We could make this test a bit clearer about what it means that the document is already present.

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapper.java (hunk 1 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapper.java]: ok for separate PR?
perhaps `SupplementaryEvidenceBuilder`?

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapperTest.java (hunk 3 above)
Comment (rkondratowicz) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapperTest.java]: done

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapper.java (hunk 1 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapper.java]: shouldn't this be in `DocumentMapper` instead? judging from arguments

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapper.java [hunk 4]
```diff
@@ -13,7 +18,20 @@ public SupplementaryEvidenceMapper() {
...
+
+        Stream<Document> docsToAdd =
+            envelopeDocs
+                .stream()
+                .filter(d -> existingDocs.stream().noneMatch(e -> Objects.equals(e.url, d.url)));
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapper.java]: should the check for dcn be here too?

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapper.java (hunk 1 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapper.java]: it's creating an object of type `SupplementaryEvidence`

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapper.java (hunk 4 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapper.java]: just the url, it uniquely identifies the file

Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/SupplementaryEvidenceMapper.java]: ah, the description in the ticket now have changed

**Pull request merged**